File: backend/services/nasa_service.py
"""
NASA API Service — async calls to NASA Exoplanet Archive and NeoWs.
All results are cached in-memory to avoid hitting rate limits.
"""
from __future__ import annotations

import asyncio
import time
from typing import Any, Dict, List, Optional, Tuple
import httpx

from models.schemas import (
    AsteroidCloseApproach,
    AsteroidData,
    AsteroidResponse,
    CelestialObject,
    ExoplanetData,
    ExoplanetResponse,
)


# Simple in-memory cache: key → (data, timestamp)
_cache: Dict[str, Tuple[Any, float]] = {}
CACHE_TTL_SECONDS = 3600  # 1 hour


def _get_cached(key: str) -> Optional[Any]:
    if key in _cache:
        data, ts = _cache[key]
        if time.time() - ts < CACHE_TTL_SECONDS:
            return data
    return None


def _set_cached(key: str, data: Any) -> None:
    _cache[key] = (data, time.time())


# NASA Exoplanet Archive TAP endpoint
EXOPLANET_BASE = "https://exoplanetarchive.ipac.caltech.edu/TAP/sync"

# NASA NeoWs (Near Earth Objects)
NEOWS_BASE = "https://api.nasa.gov/neo/rest/v1"


async def fetch_exoplanets(
    limit: int = 50,
    habitable_only: bool = False,
    api_key: str = "DEMO_KEY",
) -> List[ExoplanetResponse]:
    """Query the NASA Exoplanet Archive for confirmed exoplanets."""
    cache_key = f"exoplanets:{limit}:{habitable_only}"
    cached = _get_cached(cache_key)
    if cached is not None:
        return cached

    # ADQL query for exoplanet data
    if habitable_only:
        # Habitable zone rough estimate: equilibrium temp 200-350K, radius 0.5-2.0 Earth radii
        where_clause = (
            "default_flag=1 AND pl_rade IS NOT NULL "
            "AND pl_rade < 2.0 AND pl_rade > 0.5 "
            "AND pl_eqt IS NOT NULL AND pl_eqt > 200 AND pl_eqt < 350"
        )
    else:
        where_clause = "default_flag=1 AND pl_rade IS NOT NULL"

    query = (
        f"SELECT pl_name, pl_rade, pl_masse, pl_orbper, pl_orbsmax, "
        f"st_teff, st_rad, sy_dist, ra, dec, pl_eqt "
        f"FROM ps WHERE {where_clause} "
        f"ORDER BY sy_dist ASC "
        f"LIMIT {limit}"
    )

    params = {"query": query, "format": "json"}

    async with httpx.AsyncClient(timeout=30.0) as client:
        try:
            resp = await client.get(EXOPLANET_BASE, params=params)
            resp.raise_for_status()
            raw_data = resp.json()
        except httpx.TimeoutException:
            raise RuntimeError("NASA Exoplanet Archive timed out")
        except httpx.HTTPStatusError as e:
            raise RuntimeError(f"NASA Exoplanet Archive returned {e.response.status_code}")

    results: List[ExoplanetResponse] = []
    for row in raw_data:
        name = row.get("pl_name", "Unknown")
        # Convert distance from parsecs to light-years
        dist_parsec = row.get("sy_dist")
        dist_ly = dist_parsec * 3.26156 if dist_parsec is not None else None

        star_temp = row.get("st_teff")
        eq_temp = row.get("pl_eqt")
        radius = row.get("pl_rade")

        is_habitable = (
            eq_temp is not None
            and 200 <= eq_temp <= 350
            and radius is not None
            and 0.5 <= radius <= 2.0
        )

        exo_data = ExoplanetData(
            radius_earth=radius,
            mass_earth=row.get("pl_masse"),
            orbital_period_days=row.get("pl_orbper"),
            semi_major_axis_au=row.get("pl_orbsmax"),
            star_temperature_k=star_temp,
            star_radius_solar=row.get("st_rad"),
            equilibrium_temp_k=eq_temp,
            is_habitable=is_habitable,
        )

        planet = ExoplanetResponse(
            id=name.lower().replace(" ", "_").replace("-", "_"),
            name=name,
            type="exoplanet",
            ra=row.get("ra"),
            dec=row.get("dec"),
            distance_ly=dist_ly,
            description=_describe_exoplanet(exo_data),
            exoplanet=exo_data,
        )
        results.append(planet)

    _set_cached(cache_key, results)
    return results


async def fetch_exoplanet_by_name(name: str, api_key: str = "DEMO_KEY") -> Optional[ExoplanetResponse]:
    """Fetch a single exoplanet by name."""
    # Try to find in our cached full list first
    all_planets = await fetch_exoplanets(limit=500, api_key=api_key)
    name_lower = name.lower().strip()
    for planet in all_planets:
        if planet.name.lower() == name_lower or planet.id == name_lower.replace(" ", "_"):
            return planet

    # Direct query if not in cache
    query = (
        f"SELECT pl_name, pl_rade, pl_masse, pl_orbper, pl_orbsmax, "
        f"st_teff, st_rad, sy_dist, ra, dec, pl_eqt "
        f"FROM ps WHERE default_flag=1 AND LOWER(pl_name) LIKE '%{name_lower}%' LIMIT 1"
    )
    params = {"query": query, "format": "json"}

    async with httpx.AsyncClient(timeout=30.0) as client:
        try:
            resp = await client.get(EXOPLANET_BASE, params=params)
            resp.raise_for_status()
            raw_data = resp.json()
        except Exception:
            return None

    if not raw_data:
        return None

    row = raw_data[0]
    name_found = row.get("pl_name", "Unknown")
    dist_parsec = row.get("sy_dist")
    dist_ly = dist_parsec * 3.26156 if dist_parsec is not None else None

    exo_data = ExoplanetData(
        radius_earth=row.get("pl_rade"),
        mass_earth=row.get("pl_masse"),
        orbital_period_days=row.get("pl_orbper"),
        semi_major_axis_au=row.get("pl_orbsmax"),
        star_temperature_k=row.get("st_teff"),
        star_radius_solar=row.get("st_rad"),
        equilibrium_temp_k=row.get("pl_eqt"),
    )

    return ExoplanetResponse(
        id=name_found.lower().replace(" ", "_").replace("-", "_"),
        name=name_found,
        type="exoplanet",
        ra=row.get("ra"),
        dec=row.get("dec"),
        distance_ly=dist_ly,
        description=_describe_exoplanet(exo_data),
        exoplanet=exo_data,
    )


async def fetch_near_earth_asteroids(
    start_date: str,
    end_date: str,
    api_key: str = "DEMO_KEY",
) -> List[AsteroidResponse]:
    """Fetch near-Earth asteroids from NASA NeoWs API."""
    cache_key = f"asteroids:{start_date}:{end_date}"
    cached = _get_cached(cache_key)
    if cached is not None:
        return cached

    url = f"{NEOWS_BASE}/feed"
    params = {"start_date": start_date, "end_date": end_date, "api_key": api_key}

    async with httpx.AsyncClient(timeout=30.0) as client:
        try:
            resp = await client.get(url, params=params)
            resp.raise_for_status()
            data = resp.json()
        except httpx.TimeoutException:
            raise RuntimeError("NASA NeoWs timed out")
        except httpx.HTTPStatusError as e:
            raise RuntimeError(f"NASA NeoWs returned {e.response.status_code}")

    results: List[AsteroidResponse] = []
    near_earth_objects = data.get("near_earth_objects", {})

    for date_key, asteroid_list in near_earth_objects.items():
        for ast in asteroid_list:
            name = ast.get("name", "Unknown").strip("()")
            diameter_data = ast.get("estimated_diameter", {}).get("kilometers", {})
            close_approaches = ast.get("close_approach_data", [])

            approach: Optional[AsteroidCloseApproach] = None
            if close_approaches:
                ca = close_approaches[0]
                miss_km = float(ca.get("miss_distance", {}).get("kilometers", 0))
                rel_vel = float(ca.get("relative_velocity", {}).get("kilometers_per_second", 0))
                miss_au = float(ca.get("miss_distance", {}).get("astronomical", 0))
                approach = AsteroidCloseApproach(
                    date=ca.get("close_approach_date", date_key),
                    relative_velocity_km_s=rel_vel,
                    miss_distance_km=miss_km,
                    miss_distance_au=miss_au,
                )

            asteroid_data = AsteroidData(
                diameter_min_km=diameter_data.get("estimated_diameter_min"),
                diameter_max_km=diameter_data.get("estimated_diameter_max"),
                is_potentially_hazardous=ast.get("is_potentially_hazardous_asteroid", False),
                absolute_magnitude=ast.get("absolute_magnitude_h"),
            )

            dist_au = approach.miss_distance_au if approach else None

            asteroid = AsteroidResponse(
                id=str(ast.get("id", name.lower().replace(" ", "_"))),
                name=name,
                type="asteroid",
                distance_au=dist_au,
                description=f"Near-Earth asteroid making a close approach on {date_key}.",
                asteroid=asteroid_data,
                close_approach=approach,
            )
            results.append(asteroid)

    _set_cached(cache_key, results)
    return results


def _describe_exoplanet(data: ExoplanetData) -> str:
    """Generate a human-readable description for an exoplanet."""
    parts = []
    if data.radius_earth is not None:
        if data.radius_earth < 0.8:
            parts.append("a sub-Earth sized world")
        elif data.radius_earth <= 1.5:
            parts.append("an Earth-sized rocky world")
        elif data.radius_earth <= 4.0:
            parts.append("a super-Earth or mini-Neptune")
        else:
            parts.append("a gas giant")

    if data.equilibrium_temp_k is not None:
        temp = data.equilibrium_temp_k
        if temp < 200:
            parts.append("with a frigid frozen surface")
        elif temp <= 350:
            parts.append("potentially in the habitable zone")
        elif temp <= 700:
            parts.append("too hot for liquid water")
        else:
            parts.append("with a scorching hot surface")

    if data.star_temperature_k is not None:
        parts.append(f"orbiting a star at {data.star_temperature_k:.0f}K")

    return (". It is ".join(parts) + ".").capitalize() if parts else "An exoplanet beyond our solar system."

File: app/src/services/nasaService.js
/**
 * NASA API Service — calls to NASA Exoplanet Archive and NeoWs.
 * All results are cached in-memory to avoid hitting rate limits.
 *
 * Port of backend/services/nasa_service.py
 */

import { getCache, setCache } from './cache.js'

const CACHE_TTL = 3600 // 1 hour

// NASA Exoplanet Archive TAP endpoint
const EXOPLANET_BASE = 'https://exoplanetarchive.ipac.caltech.edu/TAP/sync'

// NASA NeoWs (Near Earth Objects)
const NEOWS_BASE = 'https://api.nasa.gov/neo/rest/v1'

// ── Exoplanets ──────────────────────────────────────────────────────────────

/**
 * Query the NASA Exoplanet Archive for confirmed exoplanets.
 * @param {number} limit
 * @param {boolean} habitableOnly
 * @param {string} apiKey
 * @returns {Promise<import('../types/astronomy.js').ExoplanetResponse[]>}
 */
export async function fetchExoplanets(limit = 50, habitableOnly = false, apiKey = 'DEMO_KEY') {
  const cacheKey = `exoplanets:${limit}:${habitableOnly}`
  const cached = getCache(cacheKey, CACHE_TTL)
  if (cached !== null) return cached

  let whereClause
  if (habitableOnly) {
    whereClause =
      'default_flag=1 AND pl_rade IS NOT NULL ' +
      'AND pl_rade < 2.0 AND pl_rade > 0.5 ' +
      'AND pl_eqt IS NOT NULL AND pl_eqt > 200 AND pl_eqt < 350'
  } else {
    whereClause = 'default_flag=1 AND pl_rade IS NOT NULL'
  }

  const query =
    `SELECT TOP ${limit} pl_name, pl_rade, pl_masse, pl_orbper, pl_orbsmax, ` +
    `st_teff, st_rad, sy_dist, ra, dec, pl_eqt ` +
    `FROM ps WHERE ${whereClause} ` +
    `ORDER BY sy_dist ASC`

  const url = new URL(EXOPLANET_BASE)
  url.searchParams.set('query', query)
  url.searchParams.set('format', 'json')

  let rawData
  try {
    const resp = await fetch(url.toString(), { signal: AbortSignal.timeout(30000) })
    if (!resp.ok) throw new Error(`NASA Exoplanet Archive returned ${resp.status}`)
    rawData = await resp.json()
  } catch (err) {
    if (err.name === 'TimeoutError') throw new Error('NASA Exoplanet Archive timed out')
    throw new Error(`NASA Exoplanet Archive error: ${err.message}`)
  }

  const results = rawData.map((row) => {
    const distParsec = row.sy_dist ?? null
    const distLy = distParsec !== null ? distParsec * 3.26156 : null
    const radius = row.pl_rade ?? null
    const eqTemp = row.pl_eqt ?? null
    const starTemp = row.st_teff ?? null

    const isHabitable =
      eqTemp !== null && eqTemp >= 200 && eqTemp <= 350 &&
      radius !== null && radius >= 0.5 && radius <= 2.0

    /** @type {import('../types/astronomy.js').ExoplanetData} */
    const exoData = {
      radius_earth: radius,
      mass_earth: row.pl_masse ?? null,
      orbital_period_days: row.pl_orbper ?? null,
      semi_major_axis_au: row.pl_orbsmax ?? null,
      star_temperature_k: starTemp,
      star_radius_solar: row.st_rad ?? null,
      equilibrium_temp_k: eqTemp,
      is_habitable: isHabitable,
    }

    const name = row.pl_name ?? 'Unknown'
    return {
      id: name.toLowerCase().replace(/ /g, '_').replace(/-/g, '_'),
      name,
      type: 'exoplanet',
      ra: row.ra ?? null,
      dec: row.dec ?? null,
      distance_ly: distLy,
      description: _describeExoplanet(exoData),
      exoplanet: exoData,
    }
  })

  setCache(cacheKey, results)
  return results
}

/**
 * Fetch a single exoplanet by name.
 * @param {string} name
 * @param {string} apiKey
 * @returns {Promise<import('../types/astronomy.js').ExoplanetResponse|null>}
 */
export async function fetchExoplanetByName(name, apiKey = 'DEMO_KEY') {
  // Try the cached full list first
  const allPlanets = await fetchExoplanets(500, false, apiKey)
  const nameLower = name.toLowerCase().trim()
  for (const planet of allPlanets) {
    if (
      planet.name.toLowerCase() === nameLower ||
      planet.id === nameLower.replace(/ /g, '_')
    ) {
      return planet
    }
  }

  // Direct query fallback
  const query =
    `SELECT TOP 1 pl_name, pl_rade, pl_masse, pl_orbper, pl_orbsmax, ` +
    `st_teff, st_rad, sy_dist, ra, dec, pl_eqt ` +
    `FROM ps WHERE default_flag=1 AND LOWER(pl_name) LIKE '%${nameLower}%'`

  const url = new URL(EXOPLANET_BASE)
  url.searchParams.set('query', query)
  url.searchParams.set('format', 'json')

  try {
    const resp = await fetch(url.toString(), { signal: AbortSignal.timeout(30000) })
    if (!resp.ok) return null
    const rawData = await resp.json()
    if (!rawData.length) return null
    return _buildExoplanetFromRow(rawData[0])
  } catch {
    return null
  }
}

/**
 * Find a confirmed exoplanet whose host star lies within radiusDeg of the given coords.
 * @param {number} ra
 * @param {number} dec
 * @param {number} radiusDeg
 * @param {string} apiKey
 * @returns {Promise<import('../types/astronomy.js').ExoplanetResponse|null>}
 */
export async function fetchExoplanetByCoordinates(ra, dec, radiusDeg = 0.05, apiKey = 'DEMO_KEY') {
  const cacheKey = `exo_coord:${ra.toFixed(3)}:${dec.toFixed(3)}:${radiusDeg}`
  const cached = getCache(cacheKey, CACHE_TTL)
  if (cached !== null) return cached

  const query =
    `SELECT TOP 1 pl_name, pl_rade, pl_masse, pl_orbper, pl_orbsmax, ` +
    `st_teff, st_rad, sy_dist, ra, dec, pl_eqt ` +
    `FROM ps WHERE default_flag=1 ` +
    `AND CONTAINS(POINT('ICRS', ra, dec), CIRCLE('ICRS', ${ra}, ${dec}, ${radiusDeg})) = 1`

  const url = new URL(EXOPLANET_BASE)
  url.searchParams.set('query', query)
  url.searchParams.set('format', 'json')

  try {
    const resp = await fetch(url.toString(), { signal: AbortSignal.timeout(30000) })
    if (!resp.ok) return null
    const rawData = await resp.json()
    if (!rawData.length) return null
    const result = _buildExoplanetFromRow(rawData[0])
    setCache(cacheKey, result)
    return result
  } catch {
    return null
  }
}

// ── Asteroids ───────────────────────────────────────────────────────────────

/**
 * Fetch near-Earth asteroids from NASA NeoWs API.
 * @param {string} startDate  YYYY-MM-DD
 * @param {string} endDate    YYYY-MM-DD
 * @param {string} apiKey
 * @returns {Promise<import('../types/astronomy.js').AsteroidResponse[]>}
 */
export async function fetchNearEarthAsteroids(startDate, endDate, apiKey = 'DEMO_KEY') {
  const cacheKey = `asteroids:${startDate}:${endDate}`
  const cached = getCache(cacheKey, CACHE_TTL)
  if (cached !== null) return cached

  const url = new URL(`${NEOWS_BASE}/feed`)
  url.searchParams.set('start_date', startDate)
  url.searchParams.set('end_date', endDate)
  url.searchParams.set('api_key', apiKey)

  let data
  try {
    const resp = await fetch(url.toString(), { signal: AbortSignal.timeout(30000) })
    if (!resp.ok) throw new Error(`NASA NeoWs returned ${resp.status}`)
    data = await resp.json()
  } catch (err) {
    if (err.name === 'TimeoutError') throw new Error('NASA NeoWs timed out')
    throw new Error(`NASA NeoWs error: ${err.message}`)
  }

  /** @type {import('../types/astronomy.js').AsteroidResponse[]} */
  const results = []
  const nearEarthObjects = data.near_earth_objects ?? {}

  for (const [dateKey, asteroidList] of Object.entries(nearEarthObjects)) {
    for (const ast of asteroidList) {
      const name = (ast.name ?? 'Unknown').replace(/^[()]+|[()]+$/g, '').trim()
      const diameterData = ast.estimated_diameter?.kilometers ?? {}
      const closeApproaches = ast.close_approach_data ?? []

      let approach = null
      if (closeApproaches.length > 0) {
        const ca = closeApproaches[0]
        approach = {
          date: ca.close_approach_date ?? dateKey,
          relative_velocity_km_s: parseFloat(ca.relative_velocity?.kilometers_per_second ?? 0),
          miss_distance_km: parseFloat(ca.miss_distance?.kilometers ?? 0),
          miss_distance_au: parseFloat(ca.miss_distance?.astronomical ?? 0),
        }
      }

      const asteroidData = {
        diameter_min_km: diameterData.estimated_diameter_min ?? null,
        diameter_max_km: diameterData.estimated_diameter_max ?? null,
        is_potentially_hazardous: ast.is_potentially_hazardous_asteroid ?? false,
        absolute_magnitude: ast.absolute_magnitude_h ?? null,
      }

      results.push({
        id: String(ast.id ?? name.toLowerCase().replace(/ /g, '_')),
        name,
        type: 'asteroid',
        distance_au: approach?.miss_distance_au ?? null,
        description: `Near-Earth asteroid making a close approach on ${dateKey}.`,
        asteroid: asteroidData,
        close_approach: approach,
      })
    }
  }

  setCache(cacheKey, results)
  return results
}

// ── Private helpers ─────────────────────────────────────────────────────────

/**
 * Build an ExoplanetResponse from a single NASA Exoplanet Archive `ps` row.
 * @param {Record<string, any>} row
 * @returns {import('../types/astronomy.js').ExoplanetResponse}
 */
function _buildExoplanetFromRow(row) {
  const name = row.pl_name ?? 'Unknown'
  const distParsec = row.sy_dist ?? null
  const distLy = distParsec !== null ? distParsec * 3.26156 : null

  const exoData = {
    radius_earth: row.pl_rade ?? null,
    mass_earth: row.pl_masse ?? null,
    orbital_period_days: row.pl_orbper ?? null,
    semi_major_axis_au: row.pl_orbsmax ?? null,
    star_temperature_k: row.st_teff ?? null,
    star_radius_solar: row.st_rad ?? null,
    equilibrium_temp_k: row.pl_eqt ?? null,
    is_habitable: false,
  }

  return {
    id: name.toLowerCase().replace(/ /g, '_').replace(/-/g, '_'),
    name,
    type: 'exoplanet',
    ra: row.ra ?? null,
    dec: row.dec ?? null,
    distance_ly: distLy,
    description: _describeExoplanet(exoData),
    exoplanet: exoData,
  }
}

/**
 * Generate a human-readable description for an exoplanet.
 * @param {import('../types/astronomy.js').ExoplanetData} data
 * @returns {string}
 */
function _describeExoplanet(data) {
  const parts = []

  if (data.radius_earth !== null) {
    if (data.radius_earth < 0.8) parts.push('a sub-Earth sized world')
    else if (data.radius_earth <= 1.5) parts.push('an Earth-sized rocky world')
    else if (data.radius_earth <= 4.0) parts.push('a super-Earth or mini-Neptune')
    else parts.push('a gas giant')
  }

  if (data.equilibrium_temp_k !== null) {
    const temp = data.equilibrium_temp_k
    if (temp < 200) parts.push('with a frigid frozen surface')
    else if (temp <= 350) parts.push('potentially in the habitable zone')
    else if (temp <= 700) parts.push('too hot for liquid water')
    else parts.push('with a scorching hot surface')
  }

  if (data.star_temperature_k !== null) {
    parts.push(`orbiting a star at ${Math.round(data.star_temperature_k)}K`)
  }

  if (!parts.length) return 'An exoplanet beyond our solar system.'
  const joined = parts.join('. It is ')
  return joined.charAt(0).toUpperCase() + joined.slice(1) + '.'
}
